### PR TITLE
bench: update nanobench add `-min_time`

### PR DESCRIPTION
**Pull request by martinus**
This PR updates the nanobench with the latest release from upstream, v4.3.6. It fixes the missing performance counters.

Due to discussions on #22999 I have done some work that should make the benchmark results more reliable. It introduces a new flag `-min_time` that allows to run a benchmark for much longer then the default. When results are unreliable, choosing a large timeframe here should usually get repeatable results even when frequency scaling cannot be disabled. The default is now 10ms. For this to work I have changed the `AddrManGood` and `EvictionProtection` benchmarks so they work with any number of iterations.

Also, this adds more usage documentation to `bench_bitcoin -h` and I've cherry-picked two changes from #22999 authored by Jon Atack

### CODE REVIEW

**Review by jonatack — COMMENTED** [review]
Nice--thank you for working on this! Using the min duration arg does improve stability for me (lower err%) and is convenient as a runtime option.

Tested Concept/Approach ACK with various suggestions below.

File: src/bench/bench.cpp [hunk 1]
```diff
@@ -61,6 +59,11 @@ void benchmark::BenchRunner::RunAll(const Args& args)
 
         Bench bench;
         bench.name(p.first);
+        if (args.min_duration > std::chrono::milliseconds::zero()) {
+            // convert to nanos before dividing to reduce rounding errors
+            bench.minEpochTime(std::chrono::nanoseconds(args.min_duration) / bench.epochs());
```
Comment (jonatack) [inline src/bench/bench.cpp:64]: Looking at https://en.cppreference.com/w/cpp/chrono/duration/duration_cast, would something like this be a good idea?

```cpp
// converting integral duration to integral duration of shorter divisible time unit:
    // no duration_cast needed
    std::chrono::duration<long, std::micro> int_usec = int_ms;
```

also line 62, can shorten if preferred
```diff
-        if (args.min_duration > std::chrono::milliseconds::zero()) {
+        if (args.min_duration > 0ms) {
```

also, where `std::chrono` is added, look at include headers for `<chrono>` where needed

File: src/bench/bench_bitcoin.cpp [hunk 2]
```diff
@@ -48,17 +51,63 @@ int main(int argc, char** argv)
...
+                     "  done with the environment variable NANOBENCH_ENDLESS. E.g. like so:\n"
+                     "\n"
+                     "    NANOBENCH_ENDLESS=MuHash ./bench_bitcoin -filter=MuHash\n"
+                     "\n"
+                     "  In rare cases it can be useful to suppress stability warnings. This can be\n"
+                     "  done with the environment variable NANOBENCH_SUPPRESS_WARNINGS, e.g:\n"
+                     "\n"
+                     "    NANOBENCH_SUPPRESS_WARNINGS=1 ./bench_bitcoin\n"
+                     "\n"
+                  << argsman.GetHelpMessage()
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/bench/bench_bitcoin.cpp:88]: Another alternative would be to add this Description and Notes documentation to `doc/benchmarking.md`.

File: src/bench/peer_eviction.cpp [hunk 3]
```diff
@@ -20,19 +20,17 @@ static void EvictionProtectionCommon(
...
     }
 
-    std::vector<Candidates> copies{
-        static_cast<size_t>(bench.epochs() * bench.epochIterations()), candidates};
-    size_t i{0};
+
     bench.run([&] {
-        ProtectEvictionCandidatesByRatio(copies.at(i));
-        ++i;
+        // creating a copy has an overhead of about 3%, so it does not influcence the benchmark results much.
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/peer_eviction.cpp]: fixed in eb399ab0577dffe7e1080bb497ff642de1c8c359

File: src/bench/addrman.cpp [hunk 4]
```diff
@@ -130,10 +115,20 @@ static void AddrManGood(benchmark::Bench& bench)
...
 
-    uint64_t i = 0;
+    // the benchmark is quite slow, so we reduce the number of epochs (measurements) for faster results.
+    bench.epochs(5);
+
     bench.run([&] {
-        markSomeAsGood(*addrmans.at(i));
-        ++i;
+        // To make the benchmark independent of the number of evaluations, we always prepare a new addrman.
+        // This is necessary because CAddrMan::Good() method modifies theobject, affecting the timing of subsequent calls
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/addrman.cpp]: fixed in 652faa16a1410b51a5ad78a3121cf6ab9329f640

File: src/bench/bench_bitcoin.cpp [hunk 5]
```diff
@@ -4,21 +4,24 @@
...
 
+#include <clientversion.h>
 #include <crypto/sha256.h>
 #include <util/strencodings.h>
 #include <util/system.h>
 
 #include <memory>
 
 static const char* DEFAULT_BENCH_FILTER = ".*";
+static const int64_t DEFAULT_MIN_DURATION_MS = 10;
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/bench_bitcoin.cpp]: fixed in b8f16c323e02c3d06e5ad57b6c0f96196d9b12e0

File: src/bench/bench_bitcoin.cpp [hunk 6]
```diff
@@ -48,17 +51,63 @@ int main(int argc, char** argv)
...
+                     "\n"
+                     "  * Use pyperf [1] to disable frequency scaling, turbo boost etc. For best\n"
+                     "    use CPU pinning and CPU isolation (see [2]).\n"
+                     "\n"
+                     "  * Each call of run() should do exactly the same work. E.g. inserting into\n"
+                     "    a std::vector doesn't do that as it will reallocate on certain calls. Make\n"
+                     "    sure each run has exactly the same preconditions.\n"
+                     "\n"
+                     "  * If results are still not reliable increase runtime with e.g.\n"
+                     "    -min_duration_ms=5000 to let a benchmark run for at least 5 seconds\n"
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/bench_bitcoin.cpp]: wordings fixed in 57a7a020a87f13c231b44ad1861b0bb6e6f92e70

File: src/bench/bench.h [hunk 7]
```diff
@@ -41,11 +41,12 @@ using ankerl::nanobench::Bench;
 typedef std::function<void(Bench&)> BenchFunction;
 
 struct Args {
-    std::string regex_filter;
     bool is_list_only;
     std::vector<double> asymptote;
     std::string output_csv;
     std::string output_json;
+    std::string regex_filter;
+    std::chrono::milliseconds min_duration;
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/bench.h]: done!

File: src/bench/rollingbloom.cpp [hunk 8]
```diff
@@ -13,16 +13,16 @@ static void RollingBloom(benchmark::Bench& bench)
...
         filter.insert(data);
 
-        data[0] = count >> 24;
-        data[1] = count >> 16;
-        data[2] = count >> 8;
-        data[3] = count;
+        data[0] = static_cast<unsigned char>(count >> 24);
+        data[1] = static_cast<unsigned char>(count >> 16);
+        data[2] = static_cast<unsigned char>(count >> 8);
+        data[3] = static_cast<unsigned char>(count);
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/rollingbloom.cpp]: I can't easily change that because the RollingBloomHash's API uses that and I don't want the change to be too large. I've changed this though to use `& 0xFF` like it's done in many other places in the code. Done in b901174bfb6895c8e332c2b5e280348836398e68

File: src/bench/bench_bitcoin.cpp [hunk 9]
```diff
@@ -4,21 +4,24 @@
...
 static void SetupBenchArgs(ArgsManager& argsman)
 {
     SetupHelpOptions(argsman);
 
-    argsman.AddArg("-asymptote=n1,n2,n3,...", "Test asymptotic growth of the runtime of an algorithm, if supported by the benchmark", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-asymptote=<n1,n2,n3,...>", "Test asymptotic growth of the runtime of an algorithm, if supported by the benchmark", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-filter=<regex>", strprintf("Regular expression filter to select benchmark by name (default: %s)", DEFAULT_BENCH_FILTER), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-list", "List benchmarks without executing them", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-list", "List benchmarks without executing them", ArgsManager::ALLOW_BOOL, OptionsCategory::OPTIONS);
+    argsman.AddArg("-min_duration_ms=<milliseconds>", strprintf("Minimum runtime per benchmark, in milliseconds (default: %d)", DEFAULT_MIN_DURATION_MS), ArgsManager::ALLOW_INT, OptionsCategory::OPTIONS);
```
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/bench_bitcoin.cpp]: done in b8f16c323e02c3d06e5ad57b6c0f96196d9b12e0

File: src/bench/bench_bitcoin.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinus) [reply src/bench/bench_bitcoin.cpp]: changed in 57a7a020a87f13c231b44ad1861b0bb6e6f92e70

File: src/bench/bench.cpp (hunk 1 above)
Comment (martinus) [reply src/bench/bench.cpp]: hm, I think it's fine to use `std::chrono::nanoseconds`, it's basically the same but with `int64_t`. It should be large so there's no easy chance for overflow. I generally try to not use `duration_cast`, because that's a sign that at that place there will be some loss of information due to the casting.

I've changed  line 64 to 0ms in b8f16c323e02c3d06e5ad57b6c0f96196d9b12e0

**Comment by martinus** [comment]
Thanks for the reviews! I've incorporated all the suggestions into the right commits and rebased everything.

File: src/bench/bench_bitcoin.cpp [hunk 10]
```diff
@@ -48,17 +51,62 @@ int main(int argc, char** argv)
...
+                     "\n"
+                     "  * Use pyperf [1] to disable frequency scaling, turbo boost etc. For best\n"
+                     "    results, use CPU pinning and CPU isolation (see [2]).\n"
+                     "\n"
+                     "  * Each call of run() should do exactly the same work. E.g. inserting into\n"
+                     "    a std::vector doesn't do that as it will reallocate on certain calls. Make\n"
+                     "    sure each run has exactly the same preconditions.\n"
+                     "\n"
+                     "  * If results are still not reliable, increase runtime with e.g.\n"
+                     "    -min_duration_ms=5000 to let a benchmark run for at least 5 seconds.\n"
```
Comment (jonatack) [inline src/bench/bench_bitcoin.cpp:71]: 57a7a02
```suggestion
                     "    -min_time=5000 to let a benchmark run for at least 5 seconds.\n"
```

File: src/bench/bench.cpp [hunk 11]
```diff
@@ -61,6 +59,12 @@ void benchmark::BenchRunner::RunAll(const Args& args)
 
         Bench bench;
         bench.name(p.first);
+        if (args.min_time > 0ms) {
+            // convert to nanos before dividing to reduce rounding errors
+            std::chrono::nanoseconds min_time_ns = args.min_time;
```
Comment (jonatack) [inline src/bench/bench.cpp:64]: b8f16c3

- per `doc/developer-notes.md`, it looks like `#include <chrono>` should be added to bench.cpp and bench_bitcoin.cpp:
```
- Every `.cpp` and `.h` file should `#include` every header file it directly uses classes, functions or other
  definitions from, even if those headers are already included indirectly through other headers.

  - *Rationale*: Excluding headers because they are already indirectly included results in compilation
    failures when those indirect dependencies change. Furthermore, it obscures what the real code
    dependencies are.
```

- pico-nit, feel free to ignore
```suggestion
            const std::chrono::nanoseconds min_time_ns{args.min_time};
```

**Review by jonatack — COMMENTED** [review]
Almost-ACK b901174bfb6

File: src/bench/bench.cpp (hunk 11 above)
Comment (martinus) [reply src/bench/bench.cpp:64]: hm, that include guideline is a bit hard to follow. There are lots of other missing includes, e.g. `bench_bitcoin.cpp` would need at least chrono, cstdint, vector, sstream, string, iostream. Should I add these as well?

About the constructor, I think that is one of the cases where assignment is better: chrono's assignment is more strict than the constructor, you can't do e.g. `std::chrono::nanoseconds x = 123` but you can do `std::chrono::nanoseconds x{123}`. So with the assignment it's certain that `args.min_time` is actually another chrono type that convertible without loss of information, and not just any number that could represent anything.

File: src/bench/bench_bitcoin.cpp (hunk 10 above)
Comment (martinus) [reply src/bench/bench_bitcoin.cpp]: damn, forgot that one! Fixed in the latest rebase.

File: src/bench/bench.cpp (hunk 11 above)
Comment (martinus) [reply src/bench/bench.cpp:73]: I've now added the headers in clean up includes commit

**Comment by martinus** [comment]
Rebased with header cleanups, and signed all commits.

**Comment by jonatack** [comment]
ACK e7e890079e92e346997cbdf491b5b0717cd4bcab

File: src/bench/addrman.cpp [hunk 12]
```diff
@@ -130,10 +115,17 @@ static void AddrManGood(benchmark::Bench& bench)
...
 
-    uint64_t i = 0;
     bench.run([&] {
-        markSomeAsGood(*addrmans.at(i));
-        ++i;
+        // To make the benchmark independent of the number of evaluations, we always prepare a new addrman.
+        // This is necessary because CAddrMan::Good() method modifies the object, affecting the timing of subsequent calls
+        // to the same method and we want to do the same amount of work in every loop iteration.
+        //
+        // This has some overhead (about 8% on my machine), but that overhead is constant so improvements in
```
Comment (MarcoFalke) [inline src/bench/addrman.cpp:123]: Likely going to change with #22974

Comment (martinus) [reply src/bench/addrman.cpp:123]: hm, I don't know much about what addrman does, but currently `markSomeAsGood` in the benchmark only uses every 32nd element. Can be bumped up to call Good() on every entry, maybe in random order?

Comment (martinus) [reply src/bench/addrman.cpp]: I've rebased because #22974 is already merged, and updated the benchmark in 153e686 so it makes more sense with the much faster `Good()`. Nothing else has changed

**Comment by jonatack** [comment]
re-ACK e148a5233292d156cda76cb20afb6641fc20f25e

The Win64 CI failure looks unrelated.

```
$ NANOBENCH_SUPPRESS_WARNINGS=1 ./bench_bitcoin -filter=AddrMan*.* -min_time=20000

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      762,448,657.67 |                1.31 |    6.8% |     23.20 | `AddrManAdd`
|    1,794,860,430.00 |                0.56 |    2.5% |     19.71 | `AddrManAddThenGood`
|        9,101,993.99 |              109.87 |    7.8% |     20.56 | `AddrManGetAddr`
|           11,379.90 |           87,874.25 |    2.1% |     21.93 | `AddrManSelect`
```

If you retouch, maybe order the addrman benchmark functions alphabetically, which would be the same order as the bench result output and also place the two similar functions, AddrManAdd and AddrManAddThenGood, next to each other.

File: src/bench/addrman.cpp (hunk 12 above)
Comment (MarcoFalke) [reply src/bench/addrman.cpp]: If the overhead is now 99%, then changes in the order of -50% or +100% are not noticed. Just asking to make sure the overhead isn't too much.

Comment (martinus) [reply src/bench/addrman.cpp]: This is what I get when I run `./src/bench/bench_bitcoin -min_time=10000 -filter="AddrManAdd.*"` on my machine:

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       44,370,397.17 |               22.54 |    0.6% |  448,854,352.19 |  141,515,648.83 |  3.172 |  33,913,128.43 |    0.1% |     11.05 | `AddrManAdd`
|      114,389,015.11 |                8.74 |    0.2% |1,201,264,095.33 |  364,837,620.00 |  3.293 |  78,812,510.33 |    0.1% |     10.87 | `AddrManAddThenGood`

So with the change it's ~44ms for `AddrManAdd`, ~114ms for `AddrManAddThenGood`, leaving ~70ms for just the Good() part of the benchmark

**Comment by martinus** [comment]
@jonatack Interesting that your benchmark results are so different. Did you compile for release? I can't believe my computer is ~20 times faster than yours... I compile with `./autogen.sh && ./configure --with-incompatible-bdb CC=clang CXX=clang++`

**Comment by jonatack** [comment]
@martinus no, it's a debug build and the computer is not optimized. Re-building a non-debug build would take a long time and I'd already tested it that way above. Sorry for the weird example.

**Comment by laanwj** [comment]
Code review ACK e148a5233292d156cda76cb20afb6641fc20f25e
